pkg/icingadb/ha.go

```diff
@@ -223,103 +224,117 @@ func (h *HA) controller() {
-		rows, err := tx.QueryxContext(ctx, query, envId, "y", h.instanceId, utils.UnixMilli(time.Now().Add(-1*timeout)))
-		if err != nil {
-			cancelCtx()
-			return internal.CantPerformQuery(err, query)
-		}
-		takeover := true
-		if rows.Next() {
-			instance := &v1.IcingadbInstance{}
-			err := rows.StructScan(instance)
-			if err != nil {
-				h.logger.Errorw("Can't scan currently active instance", zap.Error(err))
-			} else {
-				if shouldLog {
-					h.logger.Infow("Another instance is active",
-						zap.String("instance_id", instance.Id.String()),
-						zap.String("environment", envId.String()),
-						"heartbeat", instance.Heartbeat,
-						zap.Duration("heartbeat_age", time.Since(instance.Heartbeat.Time())))
-				}
-				takeover = false
+	var takeover bool
+
+	err := retry.WithBackoff(
+		ctx,
+		func(ctx context.Context) error {
+			takeover = false
+
+			tx, errBegin := h.db.BeginTxx(ctx, &sql.TxOptions{Isolation: sql.LevelSerializable})
+			if errBegin != nil {
+				return errors.Wrap(errBegin, "can't start transaction")
 			}
-		}
-		_ = rows.Close()
-		i := v1.IcingadbInstance{
-			EntityWithoutChecksum: v1.EntityWithoutChecksum{
-				IdMeta: v1.IdMeta{
-					Id: h.instanceId,
-				},
-			},
-			EnvironmentMeta: v1.EnvironmentMeta{
-				EnvironmentId: envId,
-			},
-			Heartbeat:                         *t,
-			Responsible:                       types.Bool{Bool: takeover || h.responsible, Valid: true},
-			EndpointId:                        s.EndpointId,
-			Icinga2Version:                    s.Version,
-			Icinga2StartTime:                  s.ProgramStart,
-			Icinga2NotificationsEnabled:       s.NotificationsEnabled,
-			Icinga2ActiveServiceChecksEnabled: s.ActiveServiceChecksEnabled,
-			Icinga2ActiveHostChecksEnabled:    s.ActiveHostChecksEnabled,
-			Icinga2EventHandlersEnabled:       s.EventHandlersEnabled,
-			Icinga2FlapDetectionEnabled:       s.FlapDetectionEnabled,
-			Icinga2PerformanceDataEnabled:     s.PerformanceDataEnabled,
-		}
 
-		stmt, _ := h.db.BuildUpsertStmt(i)
-		_, err = tx.NamedExecContext(ctx, stmt, i)
+			query := `SELECT id, heartbeat FROM icingadb_instance` +
+				` WHERE environment_id = ? AND responsible = ? AND id != ? AND heartbeat > ?`
 
-		if err != nil {
-			cancelCtx()
-			err = internal.CantPerformQuery(err, stmt)
-			if !utils.IsDeadlock(err) {
-				h.logger.Errorw("Can't update or insert instance", zap.Error(err))
-				break
-			} else {
-				if attempt > 2 {
-					// Log with info level after third attempt
-					h.logger.Infow("Can't update or insert instance. Retrying", zap.Error(err), zap.Int("retry count", attempt))
+			rows, errQuery := tx.QueryxContext(
+				ctx, query, envId, "y", h.instanceId, utils.UnixMilli(time.Now().Add(-1*timeout)),
+			)
+			if errQuery != nil {
+				return internal.CantPerformQuery(errQuery, query)
+			}
+
+			errTakeover := func() error {
```

Comment:
What about this:

```
var instance v1.IcingadbInstance
query := `SELECT id, heartbeat FROM icingadb_instance WHERE environment_id = ? AND responsible = ? AND id != ? AND heartbeat > ?`
switch err := tx.QueryRowxContext(ctx, query, envId, "y", h.instanceId, utils.UnixMilli(time.Now().Add(-1*timeout))).StructScan(&instance); err {
case sql.ErrNoRows:
	takeover = true
case nil:
	if shouldLog {
		h.logger.Infow("Another instance is active",
			zap.String("instance_id", instance.Id.String()),
			zap.String("environment", envId.String()),
			"heartbeat", instance.Heartbeat,
			zap.Duration("heartbeat_age", time.Since(instance.Heartbeat.Time())))
	}
default:
	return internal.CantPerformQuery(err, query)
}